scipy/special/special/cephes/dd_real.h

```diff
@@ -0,0 +1,558 @@
+ *
+ * for more details.  In particular, the fix for x86 computers is borrowed
+ * from his code.
+ *
+ * Yozo Hida
+ */
+
+/*
+ * This code taken from v2.3.18 of the qd package.
+ */
+
+#pragma once
+
+#include "../config.h"
+
+#include "unity.h"
+
+namespace special {
+namespace dd_real {
+
+    namespace detail {
+
+        constexpr double __DD_SPLITTER = 134217729.0;               // = 2^27 + 1
+        constexpr double __DD_SPLIT_THRESH = 6.69692879491417e+299; // = 2^996
+
+        /*************************************************************************
+         * The basic routines taking double arguments, returning 1 (or 2) doubles
+         *************************************************************************/
+
+        /* volatile is used below to prevent aggressive optimizations which may change
+         * the result of the error calculations. These volatiles wer e included in the
+         * original C code and may perhaps still be useful, e.g. if someone compiles with
+         * --ffastmath.
+         */
+
+        /* Computes fl(a+b) and err(a+b).  Assumes |a| >= |b|. */
+        SPECFUN_HOST_DEVICE inline double quick_two_sum(double a, double b, double *err) {
+            volatile double s = a + b;
+            volatile double c = s - a;
+            *err = b - c;
+            return s;
+        }
+
+        /* Computes fl(a+b) and err(a+b).  */
+        SPECFUN_HOST_DEVICE inline double two_sum(double a, double b, double *err) {
+            volatile double s = a + b;
+            volatile double c = s - a;
+            volatile double d = b - c;
+            volatile double e = s - c;
+            *err = (a - e) + d;
+            return s;
+        }
+
+        /* Computes fl(a*b) and err(a*b). */
+        SPECFUN_HOST_DEVICE inline double two_prod(double a, double b, double *err) {
+            volatile double p = a * b;
+            *err = std::fma(a, b, -p);
+            return p;
+        }
+
+        /* Computes fl(a*a) and err(a*a).  Faster than the above method. */
+        SPECFUN_HOST_DEVICE inline double two_sqr(double a, double *err) {
+            volatile double p = a * a;
+            *err = std::fma(a, a, -p);
+            return p;
+        }
+
+        /* Computes the nearest integer to d. */
+        SPECFUN_HOST_DEVICE inline double two_nint(double d) {
+            if (d == std::floor(d)) {
+                return d;
+            }
+            return std::floor(d + 0.5);
+        }
+
+    } // namespace detail
+
+    class DoubleDouble {
+      public:
+        double hi, lo;
+
+        DoubleDouble() : hi(0.0), lo(0.0) {}
+        DoubleDouble(double high, double low) : hi(high), lo(low) {}
+        explicit DoubleDouble(double high) : hi(high), lo(0.0) {}
+
+        SPECFUN_HOST_DEVICE inline DoubleDouble operator-() const { return DoubleDouble(-hi, -lo); }
+
+        SPECFUN_HOST_DEVICE inline DoubleDouble operator+(const double rhs) const {
```

Comment:
BTW `inline` is implied for `class` (or `struct`) member functions, no need for it here.